app/services/search/postgres/comment.rb

```diff
@@ -0,0 +1,89 @@
+module Search
+  module Postgres
+    class Comment
+      ATTRIBUTES = [
+        "articles.published",
+        "comments.body_markdown",
+        "comments.commentable_id AS commentable_id",
+        "comments.commentable_type AS commentable_type",
+        "comments.created_at",
+        "comments.id AS id",
+        "comments.public_reactions_count",
+        "comments.score",
+        "comments.user_id AS comment_user_id",
+        "podcast_episodes.podcast_id AS podcast_id",
+        "podcasts.published",
+        "users.id AS user_id",
+        "users.name",
+        "users.profile_image",
+        "users.username",
+      ].freeze
+      private_constant :ATTRIBUTES
+
+      DEFAULT_PER_PAGE = 60
+      private_constant :DEFAULT_PER_PAGE
+
+      # Because commentable represents a polymorphic relationship, ActiveRecord
+      # can't eager load the associations so we have to make all the joins
+      # manually.
+      #
+      # NOTE: if Comment::COMMENTABLE_TYPES is updated, this filter will also
+      # need to be updated
+      FORCED_EAGER_LOAD_QUERY = <<-SQL.freeze
```

Comment:
I'm open to renaming this. My thinking was the reason we're manually writing the SQL here is to "force" the eager load query because `includes` and `preload` won't work.

Reply:
This query is a bit tricky because it still results in  N + 1 queries when it's serialized. If you comment line 77, `serialize(results)` you'll notice it issues only 1 query, with or without the search term. 

But if you call `Search::Postgres::Comment.search_documents` you'll notice we're back in the N + 1 scenario. For each result, it calls `select articles.*` and `select users.*`. 

The reason being that the serializer is accessing `comment.commentable` instead of aliasing those via SQL to fields that then are joined in the final object. 

Let me try to explain with a diff, which is probably easier:

```diff
diff --git a/app/serializers/search/postgres_comment_serializer.rb b/app/serializers/search/postgres_comment_serializer.rb
index e6ba91162..de6d38e5f 100644
--- a/app/serializers/search/postgres_comment_serializer.rb
+++ b/app/serializers/search/postgres_comment_serializer.rb
@@ -3,7 +3,7 @@ module Search
   class PostgresCommentSerializer < ApplicationSerializer
     attribute :id, &:search_id
 
-    attributes :path, :public_reactions_count
+    attributes :public_reactions_count
 
     attribute :body_text, &:body_markdown
 
@@ -25,25 +25,16 @@ class PostgresCommentSerializer < ApplicationSerializer
     end
 
     attribute :hotness_score, &:score
-    attribute :published do |comment|
-      comment.commentable&.published
-    end
+    attribute :published, &:commentable_published
     attribute :published_at, &:created_at
     attribute :readable_publish_date_string, &:readable_publish_date
-    attribute :title do |comment|
-      comment.commentable&.title
-    end
+    attribute :title, &:commentable_title
 
-    # NOTE: not using the `NestedUserSerializer` because we don't need the
-    # the `pro` flag on the frontend, and we also avoid hitting Redis to
-    # fetch the cached value
     attribute :user do |comment|
-      user = comment.user
-
       {
-        name: user.name,
-        profile_image_90: user.profile_image_90,
-        username: user.username
+        name: comment.user_name,
+        # profile_image_90: comment.user_profile_image_90,
+        username: comment.user_username
       }
     end
   end
diff --git a/app/services/search/postgres/comment.rb b/app/services/search/postgres/comment.rb
index 64229b9a1..56702175b 100644
--- a/app/services/search/postgres/comment.rb
+++ b/app/services/search/postgres/comment.rb
@@ -2,21 +2,19 @@ module Search
   module Postgres
     class Comment
       ATTRIBUTES = [
-        "articles.published",
+        "COALESCE(articles.published, podcasts.published, false) AS commentable_published",
+        "COALESCE(articles.title, podcasts.title, '') AS commentable_title",
         "comments.body_markdown",
-        "comments.commentable_id AS commentable_id",
-        "comments.commentable_type AS commentable_type",
+        "comments.commentable_id",
+        "comments.commentable_type",
         "comments.created_at",
         "comments.id AS id",
         "comments.public_reactions_count",
         "comments.score",
-        "comments.user_id AS comment_user_id",
-        "podcast_episodes.podcast_id AS podcast_id",
-        "podcasts.published",
         "users.id AS user_id",
-        "users.name",
-        "users.profile_image",
-        "users.username",
+        "users.name AS user_name",
+        "users.profile_image AS user_profile_image",
+        "users.username AS user_username",
       ].freeze
       private_constant :ATTRIBUTES
 ```

As you can see I removed a few unnecessary columns and aliased selected attributes to names I then refer to in the serializer. 

Exactly because you can't "preload", the moment you access `.commentable`
 or `.user` will issue a SQL query to load that object in, causing the N + 1.

This results in one query: 

```ruby
[47] pry(main)> Search::Postgres::Comment.search_documents(per_page: 3, term: "rhymes")
  Comment Load (47.6ms)  SELECT ((ts_headline('simple', (coalesce("comments"."body_markdown"::text, '')), (to_tsquery('simple', ''' ' || 'rhymes' || ' ''' || ':*')), 'StartSel = "<mark>", StopSel = "</mark>", MaxFragments = 2'))) AS pg_search_highlight, comments.*, COALESCE(articles.published, podcasts.published, false) AS commentable_published, COALESCE(articles.title, podcasts.title, '') AS commentable_title, "comments"."body_markdown", "comments"."commentable_id", "comments"."commentable_type", "comments"."created_at", comments.id AS id, "comments"."public_reactions_count", "comments"."score", users.id AS user_id, users.name AS user_name, users.profile_image AS user_profile_image, users.username AS user_username FROM "comments" LEFT JOIN users
          ON comments.user_id = users.id
        LEFT JOIN articles
          ON comments.commentable_id = articles.id
          AND comments.commentable_type = 'Article'
        LEFT JOIN podcast_episodes
          ON comments.commentable_id = podcast_episodes.id
          AND comments.commentable_type = 'PodcastEpisode'
        LEFT JOIN podcasts
          ON podcast_episodes.podcast_id = podcasts.id INNER JOIN (SELECT "comments"."id" AS pg_search_id, (ts_rank((to_tsvector('simple', coalesce("comments"."body_markdown"::text, ''))), (to_tsquery('simple', ''' ' || 'rhymes' || ' ''' || ':*')), 0)) AS rank FROM "comments" WHERE ((to_tsvector('simple', coalesce("comments"."body_markdown"::text, ''))) @@ (to_tsquery('simple', ''' ' || 'rhymes' || ' ''' || ':*')))) AS pg_search_2b4244fb2663714ee72419 ON "comments"."id" = pg_search_2b4244fb2663714ee72419.pg_search_id WHERE (        comments.deleted = false AND
        comments.hidden_by_commentable_user = false AND
        (articles.published = true OR podcasts.published = true)
) ORDER BY "comments"."score" DESC LIMIT $1 OFFSET $2  [["LIMIT", 3], ["OFFSET", 0]]
=> [{:id=>"comment_12009",
  :public_reactions_count=>1,
....
```

I temporarily hid both `path` and `profile_image_90` from the serializer, as they'll both trigger a SQL query to the user. My suggestion is to employ a similar strategy as in 

https://github.com/forem/forem/blob/9f48b2cfb72de8cb7957e859559cf6509c407d88/app/services/search/postgres/reading_list.rb#L43-L59

By loading the users separately, you can then pass them as a parameter to the serializer. 

You'll always have 2 queries (1 for the comments and 1 for the users) but it's much better than N + 1.

This should also allow you to remove the `LEFT JOIN` on `users`